libvast/vast/type.hpp

```diff
@@ -328,6 +330,9 @@ class type final : public stateful_type_base {
   [[nodiscard]] std::optional<std::string_view>
   attribute(const char* key) && = delete;
 
+  /// Returns whether a type has any custom name.
+  [[nodiscard]] bool has_name() const noexcept;
```

Comment:
There's a gap between the mental model we convey in the documentation and what `has_name` conveys. Users think in terms of *aliases*, only those have a custom name. So I'd rather use a function `is_alias()` that implies that the type name is not a default name.

I know it doesn't exist, but the mental model would call for something like `holds_alternative<alias_type>`.

Reply:
The bigger problem is that every type has a name now, so `has_name` is confusing at best.

Reply:
>  Users think in terms of aliases

Note that this API is for developers, not for users. The mental model here is that a type can be any of the concrete types with multiple metadata layers attached.

I agree that the name is confusing, but if we rename it, then we should also rename `has_attributes`, because I'd like to not break the symmetry between the two.

Reply:
> Note that this API is for developers, not for users. 

I don't think it makes sense for the two to diverge, rather, I'd like to bring them closer together to avoid having to manage *two* mental models, and achieve a better DX. What's the problem with converging?

What's wrong with a free function to check whether a type is an alias, e.g., `is_alias(t)`? Then I know how to interpret `t.name()`.

Reply:
I also think that `is_alias()` or the inverse `is_builtin()` are more common solution to this problem.

I was wondering why this function is even needed, currently we use it in `to_legacy_type()` and `assign_metadata()`.
Only the latter is interesting long term so maybe we can even localize the function?
I also just got the idea of representing the name as the `name` attribute instead of a separate field, but I'm not sure this would give us a net benefit.

Reply:
`is_builtin` is not the inverse of `is_alias`, and that is the main source of confusion here. What we need is a function that checks whether a type has a custom name, ignoring whether the type has any attributes.

Anyways, I came up with something I think we all can agree on, and implemented and pushed that as a separate commit. Here's the gist of it:
- A new private function `is_alias` that does exactly what `type::has_name` did before, that comes with a fat exclaimer that using it leads to bad code that is unlikely to be well-designed for the type system. It is only available within `type.cpp`.
- I restructured all uses of `type::has_name` outside of `type.cpp` not to rely on the function. These weren't all that many, so I think that's fine.

> I also just got the idea of representing the name as the name attribute instead of a separate field, but I'm not sure this would give us a net benefit.

Not implementable in a backwards-compatible manner, I think. We also call `type::name` and `type::names` in the hot path, and doing a binary search is just a waste of cycles here. The code paths that look at attributes are entirely separate from the code paths that look at names.